tfx/orchestration/experimental/kubernetes/kubernetes_remote_runner.py

```diff
@@ -0,0 +1,194 @@
+          namespace='default',
+          label_selector='job-name={}'.format(pod_label)
+      ).items
+    except client.rest.ApiException as e:
+      if e.status != 404:
+        raise RuntimeError('Unknown error! \nReason: %s\nBody: %s' %
+                           (e.reason, e.body))
+    time.sleep(1)
+
+  # Transient orchestrator should only have 1 pod.
+  if len(orchestrator_pods) != 1:
+    raise RuntimeError('Expected 1 pod launched by Kubernetes job, found %d' %
+                       len(orchestrator_pods))
+  orchestrator_pod = orchestrator_pods.pop()
+  pod_name = orchestrator_pod.metadata.name
+
+  absl.logging.info('Waiting for pod "default:%s" to start.', pod_name)
+  kube_utils.wait_pod(
+      core_api,
+      pod_name,
+      'default',
+      exit_condition_lambda=kube_utils.pod_is_not_pending,
+      condition_description='non-pending status')
+
+  # Stream logs from orchestrator pod.
+  absl.logging.info('Start log streaming for pod "default:%s".', pod_name)
+  try:
+    logs = core_api.read_namespaced_pod_log(
+        name=pod_name,
+        namespace='default',
+        container=container_name,
+        follow=True,
+        _preload_content=False).stream()
+  except client.rest.ApiException as e:
+    raise RuntimeError(
+        'Failed to stream the logs from the pod!\nReason: %s\nBody: %s' %
+        (e.reason, e.body))
+
+  for log in logs:
+    absl.logging.info(log.decode().rstrip('\n'))
+
+  resp = kube_utils.wait_pod(
+      core_api,
+      pod_name,
+      'default',
+      exit_condition_lambda=kube_utils.pod_is_done,
+      condition_description='done state',
+      exponential_backoff=True)
+
+  if resp.status.phase == kube_utils.PodPhase.FAILED.value:
+    raise RuntimeError('Pod "default:%s" failed with status "%s".' %
+                       (pod_name, resp.status))
+
+
+def _serialize_pipeline(pipeline: tfx_pipeline.Pipeline) -> Text:
+  """Serializes a TFX pipeline.
+
+  To be replaced with the "portable core" of the unified TFX orchestrator:
+  https://github.com/tensorflow/community/pull/271. This serialization
+  procedure extracts from the pipeline properties necessary for reconstructing
+  the pipeline instance from within the cluster. For properties such as
+  components and metadata config that can not be directly dumped with JSON,
+  we use NodeWrapper and MessageToJson to serialize them beforehand.
+
+  Args:
+    pipeline: Logical pipeline containing pipeline args and components.
+
+  Returns:
+    Pipeline serialized as JSON string.
+  """
+  serialized_components = []
+  for component in pipeline.components:
+    serialized_components.append(
+        json_utils.dumps(node_wrapper.NodeWrapper(component)))
+  # Extract and pass pipeline graph information which are lost during the
+  # serialization process. The orchestrator container uses downstream_ids
+  # to reconstruct pipeline graph.
+  downstream_ids = json.dumps(_extract_downstream_ids(pipeline.components))
```

Comment:
It is better to not json dump this twice -- drop the json.dumps here so the structure is dumped inline with the outer dict on line 156.

Reply:
Fixed.